lib/rucio/common/utils.py

```diff
@@ -906,9 +906,7 @@ def detect_client_location():
 
     ip = '0.0.0.0'
     try:
-        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-        s.connect(("8.8.8.8", 80))
-        ip = s.getsockname()[0]
+        ip = requests.get('https://api64.ipify.org').text
```

Comment:
I see multiple problems with this: returns ipv6 address when thats the default & relies on external service

Reply:
8.8.8.8 is also external service. 
I proposed using much nicer service that we manage (https://rucio.cern.workers.dev/) but Martin wants to avoid potentially political discussions.

concerning IPv6... if you look at actual calculation of the distance only ip is used. IPv6 is used nowhere as far as I can tell. I would have completely removed ipv6 but I don't know if someone planed to use it for something.

Reply:
> 8.8.8.8 is also external service.

which is completely irrelevant when using a DGRAM socket and "connecting" to just fetch the local ip. You could use any IP address that routes to the external network here, in fact even `0.0.0.1` works for me. I agree that any option seems hacky though.

> concerning IPv6... if you look at actual calculation of the distance only ip is used. IPv6 is used nowhere as far as I can tell. I would have completely removed ipv6 but I don't know if someone planed to use it for something.

good point

Reply:
You are right, `8.8.8.8` is already external anyway, I would propose to take this out entirely and just set the ip to the one the rucio REST interface sees. We then have the option to overwrite lat/long anyway, so no need to resolve the ip locally? Not sure if I am missing something where `detect_client_location` is called otherwise?

Reply:
@bziemons problem is with socket method itself. It will give local IP if you are running in container or behind NAT.

Reply:
yes, but the server can see the client's public IP as Martin stated. I do not see a reason to have this :)

Reply:
> You are right, `8.8.8.8` is already external anyway, I would propose to take this out entirely and just set the ip to the one the rucio REST interface sees. We then have the option to overwrite lat/long anyway, so no need to resolve the ip locally? Not sure if I am missing something where `detect_client_location` is called otherwise?

I am not sure what do you want to take out. Whole function? That we can't do as it also sets Site. Or you want to replace the address contacted with some Rucio IP address?

Reply:
Only replace the ip part of the dict in the REST
E.g. https://github.com/rucio/rucio/blob/5db83564817e28207e2565f413a7857fbbba833f/lib/rucio/web/rest/flaskapi/v1/replicas.py#L292 already resolves the ip. If you replace in the `client_location` dictionary the IP there, I think that should be sufficient.

Reply:
> yes, but the server can see the client's public IP as Martin stated. I do not see a reason to have this :)

I agree that Rucio server "sees" correct client's IP address. If you know how to propagate that info to here that would be great.

Reply:
if you replace the socket.connect() with a requests.get() you will actually do a call to an outside service, we really should avoid this. the DGRAM socket connect is node-local only (since UDP has no "connection").

as a "better" alternative, I would completely move this to client-ip detection on the rucio server side.

Reply:
I think we all agree that the best option is to use address that rucio already sees. Only question is how do we reorganize things. I did the simplest thing that works and requires no restructuring.

Reply:
Have a look at the code line I posted above. There use the `client_ip` to overwrite the ip part in the `client_location` dictionary. It should be as easy as that.

Reply:
@bari12 sorry I still don't get it... detect_client_location is used at three places (rucio/list_file_replicas, downloadclient and uploadclient), non of which has access to the request object?

Reply:
this should be done now.